analyzers/src/SonarAnalyzer.Common/Rules/ArrayPassedAsParamsBase.cs

```diff
@@ -25,33 +25,44 @@ public abstract class ArrayPassedAsParamsBase<TSyntaxKind, TArgumentNode> : Sona
     where TArgumentNode : SyntaxNode
 {
     private const string DiagnosticId = "S3878";
-    protected override string MessageFormat => "Remove this array creation and simply pass the elements.";
-
-    private readonly DiagnosticDescriptor rule;
 
     protected abstract TSyntaxKind[] ExpressionKinds { get; }
     protected abstract TArgumentNode LastArgumentIfArrayCreation(SyntaxNode expression);
+    protected abstract ITypeSymbol ArrayElementType(SyntaxNode expression, SemanticModel model);
+
+    protected override string MessageFormat => "Remove this array creation and simply pass the elements.";
 
-    protected ArrayPassedAsParamsBase() : base(DiagnosticId) =>
-        rule = Language.CreateDescriptor(DiagnosticId, MessageFormat);
+    protected ArrayPassedAsParamsBase() : base(DiagnosticId) {}
 
     protected sealed override void Initialize(SonarAnalysisContext context) =>
         context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
         {
             if (LastArgumentIfArrayCreation(c.Node) is { } lastArgument
-                && ParameterSymbol(c.SemanticModel, c.Node, lastArgument) is { IsParams: true } param
+                && c.SemanticModel.GetSymbolInfo(c.Node).Symbol is IMethodSymbol methodSymbol
+                && Language.MethodParameterLookup(c.Node, methodSymbol).TryGetSymbol(lastArgument, out var param)
+                && param is { IsParams: true }
+                && !IsObjectOrArrayType(c.Node, methodSymbol, param, c.SemanticModel)
                 && !IsJaggedArrayParam(param))
             {
-                c.ReportIssue(rule, lastArgument.GetLocation());
+                c.ReportIssue(Rule, lastArgument.GetLocation());
             }
         }, ExpressionKinds);
 
-    private IParameterSymbol ParameterSymbol(SemanticModel model, SyntaxNode invocation, TArgumentNode argument) =>
-        model.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol
-        && Language.MethodParameterLookup(invocation, methodSymbol).TryGetSymbol(argument, out var param)
-            ? param
-            : null;
-
     private static bool IsJaggedArrayParam(IParameterSymbol param) =>
         param.Type is IArrayTypeSymbol { ElementType: IArrayTypeSymbol };
+
+    private bool IsObjectOrArrayType(SyntaxNode node, IMethodSymbol symbol, IParameterSymbol param, SemanticModel model)
+    {
+        return param.Type is IArrayTypeSymbol array
+            && (array.ElementType.Is(KnownType.System_Array)
+               || (array.ElementType.Is(KnownType.System_Object) && !ParamArgumentsAreReferenceTypeArrays(node, symbol, param, model)));
+
+        bool ParamArgumentsAreReferenceTypeArrays(SyntaxNode node, IMethodSymbol symbol, IParameterSymbol param, SemanticModel model) =>
+            Language.MethodParameterLookup(node, symbol) is { } parameterLookup
+            && parameterLookup.TryGetSyntax(param, out var arguments)
```

Comment:
Maybe I am confused with the flow, but aren't we getting back the same `SyntaxNode lastArgument` for which, two levels above,  we got the `param` symbol, by doing `Language.MethodParameterLookup(c.Node, methodSymbol).TryGetSymbol(lastArgument, out var param)`?
Am I missing something here?

Reply:
You are absolutely right. I was focused on making it work and missed that.
I refactored - however, I still need to pass the IMethodParameterLookup to check the parameters count during the invocation call.